lib/validators/json-schema-legacy.js

```diff
@@ -76,6 +76,10 @@ class JsonSchemaLegacy extends JsonSchemaValidator {
 
     switch (this.jsonSchemaVersion) {
       case 'draftV3':
+        // eslint-disable-line no-console
+        console.warn(
```

Comment:
I'm not sure that stdout is perhaps the right place for this. Although, I don't have any other suggestion.

Depending on how you use the library:

- On web service - it can flood the logs with this deprecated message, and it may be in the wrong format (since our services used structured log formats in JSON).
- In the web - the message is hidden in the console and usually user's are not going to see it
- Dredd -- this is one case it may actually expose the message, although it might get hidden by the test output.

Reply:
You can see this warning only in stdout and browser's console. You are right that it's not intended for indirect consumers (Apiary or Dredd users). Warning those should be handled by respective parent environment (i.e. warning message in the UI of Apiary). I don't think we should (and can) handle this on Gavel's side. 

What I think we should do is:

1. Determine amount of users using Draft V3 in production.
1. See if there is a substantial difference between drafts V3 and V4 (as you've mentioned there doesn't seem to be any).
1. If any of the preceding points give us reason to do any UI changes then we should, otherwise there is no changes that users should do.

Reply:
I'm wondering if it would make sense to put it on the validation result either inside `errors` or a new property `warnings`. But I don't know how these are consumed.

Reply:
@patricksmms it's a perfectly fine suggestion, no worries. 

When we re-designed the output of Gavel we put certain restrictions over what should be present in the validation result and `errors` in particular. Shortly, you should access those properties only if any validation has happened. Also, validation result always represents the result of matching actual HTTP message against the expected one.

Keeping that in mind, such warning is not a result of expected/actual data validation. Instead, it's a result of validating any of that data against an external factor (internally supported version of JSON Schema). We shouldn't have such messages in the validation result by design.

> It's also useful to know that Gavel provides no validation against external factors, for example that the given HTTP message is compliant with its respective specification. I'd like for it to stay that way.

Reply:
@kylef if preventing spamming the logs is our primary concern we make Gavel accept options. One of those options may be `suppressDeprecationWarnings: true`, which we can pass to Gavel calls on the Apiary side. Using Gavel directly, or under Dredd should still produce a warning evident for the developer (in stdout).

Reply:
**Okay, so to recap:**

- Gavel can be consumed directly (via its NodeJS API) and indirectly (Dredd, Apiary). 
- Deprecation message should be evident, so it's very easy to spot for the user.

Correspondingly, deprecations should be handled for both direct and indirect Gavel consumers.

## Direct consumer

I'm not aware of a better way than printing a deprecation message into `stdout`. It's also how other libraries handle deprecations, as far as I've seen.

Since Gavel can run in NodeJS indirectly, it _may_ pollute a parent's process stdout. 

## Indirect consumer

 Indirect consumer should know that some deprecation message was thrown, and propagate it in some form to the actual user. Displaying a message banner in the UI is a nice example for the client-side consumption.

Reply:
This makes sense to me.